test/integration/discovery.go

```diff
@@ -147,6 +153,124 @@ var _ = Describe("Discover Cluster", func() {
+		var deps *appsv1.Deployment = nil
+		testName := "discovery-integration-test"
+		nodeName := "kind-worker3"
+
+		BeforeEach(func() {
+			if f.Kubeconfig.CurrentContext != "kind-kind" {
+				Skip("Skipping cluster that isn't using context kind-kind")
+			}
+
+			// create a pod to test and attach to the node soon to be in a NotReady state
+			if deps == nil {
+				var err error = nil
+				deps, err = createDeployResource(kubeClient, depSingleContainerWithResources(namespace, "", 1, false, false, false, nodeName))
+				framework.ExpectNoError(err, "Error creating test resources")
+			}
+			// stop running node worker to simulate node in NotReady state
+			execute("docker", "stop", nodeName)
+			if err := wait.PollImmediate(framework.PollInterval, framework.DefaultSingleCallTimeout, func() (bool, error) {
+				node, errInternal := kubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+				if errInternal != nil {
+					glog.Errorf("Unexpected error while finding Node %s: %v", nodeName, errInternal)
+					return false, nil
+				}
+
+				return !util.NodeIsReady(node), nil
+			}); err != nil {
+				framework.Failf("Failed to put node in a NotReady state. %v", err)
+			}
+
+			entityDTOs, groupDTOs, err := discoveryClient.DiscoverWithNewFramework(testName)
+			if err != nil {
+				framework.Failf(err.Error())
+			}
+
+			entities = entityDTOs
+			groups = groupDTOs
+		})
+
+		AfterEach(func() {
+			// restart node once finished with the test
+			execute("docker", "start", nodeName)
+			if err := wait.PollImmediate(framework.PollInterval, framework.TestContext.SingleCallTimeout, func() (bool, error) {
+				node, errInternal := kubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+				if errInternal != nil {
+					glog.Errorf("Unexpected error while finding Node %s: %v", nodeName, errInternal)
+					return false, nil
+				}
+				return util.NodeIsReady(node), nil
+			}); err != nil {
+				framework.Failf("Failed to put node in a Ready state. %v", err)
+			}
+
+		})
+
+		It("unknown node should be detected", func() {
+			found := getUnknownNode(entities)
+			if found == nil {
+				framework.Failf("Node with unknown state not found")
+			}
+
+			if found.GetDisplayName() == "" {
+				framework.Failf("unknown nodes should have a node name")
+			}
+
+		})
+
+		It("unknown nodes should be listed in the unknown node group", func() {
+			unknownNode := getUnknownNode(entities)
+			unknownGroup := getUnknownNodeGroup(groups)
+
+			if unknownGroup == nil {
+				framework.Failf("Could not find Unknown Node group")
+			}
+
+			if unknownGroup.GetEntityType() != proto.EntityDTO_VIRTUAL_MACHINE {
+				framework.Failf("Unknown node group as incorrect entity type")
+			}
+
+			if unknownGroup.GetDisplayName() != "All Nodes (State Unknown)" {
+				framework.Failf("Unknown node group display name has been changed")
+			}
+
+			nodeInGroup := false
+			for _, uuid := range unknownGroup.GetMemberList().GetMember() {
```

Comment:
You can make this better by checking the size of the slice (we are expecting only 1 node here) and then checking that that one members id matches

Reply:
True, but if we keep it as is, then we would be able to handle cases where we added multiple unknown nodes. 😄